### ISSUE TITLE
file-tree not support  file-icons.....

### ISSUE THREAD

**Issue by 401KG** [facebook/nuclide#951]
there was some compatible problem between Nuclide's file-tree and 
file-icons(https://github.com/DanBrooker/file-icons)

file-icons is a very nice and friendly UI-supporting Package for Atom, 
and it was working fine with Nuclide in so much  previous verisons .

really hope both your guys can make some standard to fix it.

gl

**Comment by Alhadis** [comment]
I better weigh in on this...

Hi guys, I help maintain the File-Icons repository. I'm solely responsible for the mess I unknowingly thrust onto Nuclide's users, because I only just learned of its existence in the last 24 hours.

The reason your users are suddenly seeing [this](https://cloud.githubusercontent.com/assets/22787892/21559046/8d19c418-ce7f-11e6-922f-83719917fc25.png) in their workspace is because the [`nuclide-file-tree`](https://github.com/facebook/nuclide/blob/master/pkg/nuclide-file-tree/package.json#L7) isn't consuming Atom's icon service the way the built-in [`tree-view`](https://github.com/atom/tree-view/blob/master/package.json#L24-L30) does:

```json
"consumedServices": {
    "atom.file-icons": {
        "versions": {
            "1.0.0": "consumeFileIcons"
        }
    }
}
```

Now, before somebody rushes off to apply a hasty hotfix, I better explain the other half of the issue: Atom's icon service is too limited to be of use to File-Icons. I had to [monkey-patch five core packages](https://github.com/DanBrooker/file-icons/tree/master/lib/consumers) with absurd hackery and async gymnastics to get icons updating correctly. This was necessitated by the [Atom team's decision](https://github.com/atom/tree-view/pull/825#r72039664) to keep the service "simple": that is, barring packages any DOM access whatsoever.

Now, Atom's package code is easy to hack. Obviously, yours is not. I tried jackhammering Nuclide into cooperation with a `MutationObserver` to update element classes, which... uh... [didn't go](https://cloud.githubusercontent.com/assets/2346707/21558852/f624c894-ce94-11e6-8913-18750c085451.png) [too well](https://cloud.githubusercontent.com/assets/2346707/21558870/4e218c8a-ce95-11e6-849a-1be5fc516ec7.png).

To **TL;DR** this for you: I'm in the final stages of [developing our own icon service](https://github.com/DanBrooker/file-icons/commit/e537dacc23e798b3d3e432d90e344134e28e8d69), which will obviate gorilla-punching the crap out of package classes at runtime. **I urge you to use this first, Atom's API second**. Here's how I did it when testing it locally with a fork of `tree-view`:

```coffee
    if service = IconServices.get 'element-icons'
      service.addIconToElement @fileName,
        path: @file.path
        type: 'file'
    else
      service = IconServices.get 'file-icons'
      iconClass = service.iconClassForPath(@file.path, "tree-view")

    if iconClass
      unless Array.isArray iconClass
        iconClass = iconClass.toString().split(/\s+/g)
      @fileName.classList.add(iconClass...)
```

All that's left is to provide access to icons for directory elements, and then we're in the all clear. =)

FYI, my knowledge of React/Flow is precisely zero, as it is with any modern framework. So I'm not sure I'd be well-equipped enough to submit a PR for this...

**Comment by Alhadis** [comment]
[Integration is ready to go](https://github.com/DanBrooker/file-icons#integration-with-other-packages).

**Comment by 401KG** [comment]
up up

**Comment by nmote** [comment]
Cool, generally I'm supportive of using a first-class service-hub interface that is well-specified rather than a bunch of random DOM hackery. However I don't have any context on this particular problem. Maybe @shushz can weigh in on whether we should move forward with this?

Either way I wouldn't count on anyone on the core team taking time to work on this soon. If we decide that we'd like to move forward we will likely need community contributions to actually implement this.

**Comment by Alhadis** [comment]
Paging @jacobmischka regarding the aforementioned "community involvement" factor.

**Comment by jacobmischka** [comment]
My fork at jacobmischka/nuclide is as far as I've gotten. I'm not familiar with flow, and the current code there has some flow errors. It also needs a bit more testing and maybe some tests.

I've unfortunately not been able to get around to fixing those issues yet, but the service integration should be essentially complete. If anyone feels like forking my fork in the meantime and dealing with them and submitting a PR here that would be very welcome, otherwise I'll try to get to it this evening.

**Comment by nmote** [comment]
Keep in mind that I'm not endorsing this particular change. I don't know enough about the specifics here. I do believe that the Atom folks are generally very reasonable though, so if they had concerns about this API we may very well have the same concerns. We need somebody familiar with the file tree to weigh in here before we move forward.

**Comment by jacobmischka** [comment]
Understood, I honestly didn't plan on having a PR of mine merged, it was essentially intended as a sort of reference implementation. In that case I'll probably just be leaving my fork as is unless/until it's decided that it would be worthwhile to merge.

Thanks for keeping an open mind though!

**Comment by Alhadis** [comment]
@nmote Can you explain what you mean by this?

> I do believe that the Atom folks are generally very reasonable though, so **if they had concerns about this API we may very well have the same concerns**

**Comment by Alhadis** [comment]
I've implemented a [stopgap solution](https://github.com/file-icons/atom/commit/1d3899c86323db4181d75a953e810c268a22dbc1). Icons are still missing in some areas, and they won't update dynamically, but it's better than nothing.

I noticed that you've gone out of your way to [support our package before](https://github.com/facebook/nuclide/commit/73ad4b0a9f1d1f77b1792b592735d819f73e8483), so I'm hoping you'll do so again. =)

**Comment by Alhadis** [comment]
> Maybe @shushz can weigh in on whether we should move forward with this?

Paging @shushz in the hopes of getting this looked at.

7f09071 has me concerned Nuclide will eventually reimplement other Atom packages too, eventually introducing a change that's impossible to monkey-patch.

**Comment by shushz** [comment]
Yeah, sorry  - too many things to look at at once. Looking into it now.

**Comment by shushz** [comment]
@Alhadis so from browsing the source it seems that file-icons uses [synchronous fs](https://github.com/file-icons/atom/blob/master/lib/utils/fs.js#L8-L10) as part of its icon determination flow. With Nuclide this would be a non-starter, as one of the main focus areas for Nuclide is support for remote editing (on a remote host, that is). So all of our `fs` interactions are written with remote in mind.

Nuclide is also expected to work with trees of huge repositories, thousands of elements. For that purpose the rendering of the file-tree rendering is windowed. I suspect that adding any degree of complex logic to the rendering process would degrade performance significantly. Current approach in the file-tree is that everything that can be decided ahead of the rendering is, all data is kept in an immutable hierarchical data structure and the rendering is straightforward process from that structure.

**Comment by Daniel15** [comment]
For what it's worth, older versions seem to be working fine for me. I downgraded by running:

```
apm install file-icons@1.7.25
```

**Comment by jacobmischka** [comment]
Yeah, they will. 1.x used a global css method based on the filenames, 2.x uses a js one based on that and bunch of other metrics.

By the way, 1.7.25 is missing sourcemap icons and a couple other additions that were made before 2.0, jacobmischka/file-icons has those included as well if you care about them, or you can just look through the official repo's commits until you find it. No other 2.0 features have been backported.

**Comment by Alhadis** [comment]
> With Nuclide this would be a non-starter, as one of the main focus areas for Nuclide is support for remote editing (on a remote host, that is).

Synchronous stat-calls are only made for detecting symlinks and deduping resources based on their inode. If the stat-call fails, the resource is assumed to be either unreadable or virtual, and marked as such. In which case, the package is instructed to avoid making an attempt to scan its contents.

> Current approach in the file-tree is that everything that can be decided ahead of the rendering is, all data is kept in an immutable hierarchical data structure and the rendering is straightforward process from that structure.

If your rendering procedure is *that* restrictive, then it's better to use [Atom's simpler icon-service](https://github.com/atom/tree-view/blob/master/package.json#L24-L30). Anywhere where a `[data-path]` or `[data-name]` attribute is expected to show a file's icon should be instead using a call to the service hub.

Bottom line: the services API needs to be used one way or the other.

**Comment by jgebhardt** [comment]
This was fixed in https://github.com/facebook/nuclide/tree/release-v0.219.0.
